kernel/src/syscall.rs

```diff
@@ -69,35 +154,239 @@ impl Syscall {
+#[repr(u32)]
+#[derive(Copy, Clone, Debug)]
+pub enum SyscallReturnVariant {
+    Failure = 0,
+    FailureU32 = 1,
+    FailureU32U32 = 2,
+    FailureU64 = 3,
+    Success = 128,
+    SuccessU32 = 129,
+    SuccessU32U32 = 130,
+    SuccessU64 = 131,
+    SuccessU32U32U32 = 132,
+    SuccessU64U32 = 133,
+}
+
+/// Possible system call return variants, generic over the system call
+/// type
+///
+/// This struct operates over primitive types such as integers of
+/// fixed length and pointers. It is constructed by the scheduler and
+/// passed down to the architecture to be encoded into registers,
+/// possibly using the provided
+/// [`encode_syscall_return`](GenericSyscallReturnValue::encode_syscall_return)
+/// method.
+///
+/// Capsules use higher level Rust types
+/// (e.g. [`ReadWriteAppSlice`](crate::ReadWriteAppSlice) and
+/// [`Callback`](crate::Callback)) or wrappers around this struct
+/// ([`CommandResult`](crate::CommandResult)) which limit the
+/// available constructors to safely constructable variants.
+#[derive(Copy, Clone, Debug)]
+pub enum GenericSyscallReturnValue {
+    /// Generic error case
+    Failure(ErrorCode),
+    /// Generic error case, with an additional 32-bit data field
+    FailureU32(ErrorCode, u32),
+    /// Generic error case, with two additional 32-bit data fields
+    FailureU32U32(ErrorCode, u32, u32),
+    /// Generic error case, with an additional 64-bit data field
+    FailureU64(ErrorCode, u64),
+    /// Generic success case
+    Success,
+    /// Generic success case, with an additional 32-bit data field
+    SuccessU32(u32),
+    /// Generic success case, with two additional 32-bit data fields
+    SuccessU32U32(u32, u32),
+    /// Generic success case, with three additional 32-bit data fields
+    SuccessU32U32U32(u32, u32, u32),
+    /// Generic success case, with an additional 64-bit data field
+    SuccessU64(u64),
+    /// Generic success case, with an additional 32-bit and 64-bit
+    /// data field
+    SuccessU64U32(u64, u32),
+
+    // These following types are used by the scheduler so that it can
+    // return values to userspace in an architecture (pointer-width)
+    // independent way. The kernel passes these types (rather than
+    // AppSlice or Callback) for two reasons. First, since the
+    // kernel/scheduler makes promises about the lifetime and safety
+    // of these types (e.g., an accepted allow does not overlap with
+    // an existing accepted AppSlice), it does not want to leak them
+    // to other code. Second, if subscribe or allow calls pass invalid
+    // values (pointers out of valid memory), the kernel cannot
+    // construct an AppSlice or Callback type but needs to be able to
+    // return a failure. -pal 11/24/20
+    /// Read/Write allow success case
+    AllowReadWriteSuccess(*mut u8, usize),
+    /// Read/Write allow failure case
+    AllowReadWriteFailure(ErrorCode, *mut u8, usize),
+
+    /// Read only allow success case
```

Comment:
```suggestion
    /// Read only allow success case. Used to return
    /// the old allow buffer to the process.
```

Reply:
I'm updating this with more useful comments.